In [2]:
import pandas as pd
import numpy as np

## Pandas analysis

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a couple of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every 'x' BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [18]:
# If you didn't download it yet, please get the relevant file now!
!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ~/data/

--2021-11-24 10:50:49--  https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt
Risoluzione di www.dropbox.com (www.dropbox.com)... 2620:100:6025:18::a27d:4512, 162.125.69.18
Connessione a www.dropbox.com (www.dropbox.com)|2620:100:6025:18::a27d:4512|:443... connesso.
Richiesta HTTP inviata, in attesa di risposta... 301 Moved Permanently
Posizione: /s/raw/xvjzaxzz3ysphme/data_000637.txt [segue]
--2021-11-24 10:50:49--  https://www.dropbox.com/s/raw/xvjzaxzz3ysphme/data_000637.txt
Riutilizzo della connessione esistente a [www.dropbox.com]:443.
Richiesta HTTP inviata, in attesa di risposta... 302 Found
Posizione: https://uc9a25b6a93f8dc4383c1b6bc537.dl.dropboxusercontent.com/cd/0/inline/BalEIqVvl_L4fmBhBRB-7vdlZuKg4AsY3yr13v3XzL9kt_m0_fYRUGNPn2a0XiWuBzCsCkY6vDpqNNfUMRyytGYcqB9-pm_AofyTV0ReoaFkyuDHf6dwHSK3I1O2YiQ_dT7JmwF4F0EKKFSMTynZnDrL/file# [segue]
--2021-11-24 10:50:50--  https://uc9a25b6a93f8dc4383c1b6bc537.dl.dropboxusercontent.com/cd/0/inline/BalEIqVvl_L4fmBhBRB-7vdlZuKg4AsY3yr1

1\. Create a Pandas DataFrame reading N rows of the 'data_000637.txt' dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k.

In [3]:
#da rivedere
file_name="~/data/data_000637.txt"
data=pd.read_csv(file_name)
N=1300000
data1=data[:N]
data1

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0           1     0          123  3869200167        2374        26
1           1     0          124  3869200167        2374        27
2           1     0           63  3869200167        2553        28
3           1     0           64  3869200167        2558        19
4           1     0           64  3869200167        2760        25
...       ...   ...          ...         ...         ...       ...
1299995     1     0           62  3869211100        3077        13
1299996     1     0           49  3869211100        3081        14
1299997     1     0          139  3869211100        3084         0
1299998     1     0           59  3869211100        3085        11
1299999     1     0           63  3869211100        3078        27

[1300000 rows x 6 columns]

2\. Find out the number of BX in a ORBIT (the value 'x').

In [4]:
x=data1['BX_COUNTER'].max()+1
print(x)

3564


4\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information).

In [5]:
data['timens']=data['TDC_MEAS']*25/30+data['BX_COUNTER']*25+data['ORBIT_CNT']*x*25
data


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
0           1     0          123  3869200167        2374        26   
1           1     0          124  3869200167        2374        27   
2           1     0           63  3869200167        2553        28   
3           1     0           64  3869200167        2558        19   
4           1     0           64  3869200167        2760        25   
...       ...   ...          ...         ...         ...       ...   
1310715     1     0           62  3869211171         762        14   
1310716     1     1            4  3869211171         763        11   
1310717     1     0           64  3869211171         764         0   
1310718     1     0          139  3869211171         769         0   
1310719     1     0           61  3869211171         762        18   

               timens  
0        3.447457e+14  
1        3.447457e+14  
2        3.447457e+14  
3        3.447457e+14  
4        3.447457e+14  
...               ...  
1310715  3.447467e+14  
1310716  3.447467e+14  
1310717  3.447467e+14  
1310718  3.447467e+14  
1310719  3.447467e+14  

[1310720 rows x 7 columns]

3\. Find out how much the data taking lasted. You can either make an estimate based on the fraction of the measurements (rows) you read, or perform this check precisely by reading out the whole dataset.

In [6]:
tmax=data['timens'].max()
tmin=data['timens'].min()
print((tmax-tmin)*10**(-9))

0.9804162533125


5\. Replace the values (all 1) of the HEAD column randomly with 0 or 1.

In [ ]:
data[data]

1\. Create a Pandas DataFrame reading N rows of the 'data_000637.txt' dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k.

2\. Find out the number of BX in a ORBIT (the value 'x').

3\. Find out how much the data taking lasted. You can either make an estimate based on the fraction of the measurements (rows) you read, or perform this check precisely by reading out the whole dataset.

4\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information).

5\. Replace the values (all 1) of the HEAD column randomly with 0 or 1.

6\. Create a new DataFrame that contains only the rows with HEAD=1.

7\. Make two occupancy plots (one for each FPGA), i.e. plot the number of counts per TDC channel

8\. Use the groupby method to find out the noisy channels, i.e. the TDC channels with most counts (say the top 3)

9\. Count the number of unique orbits. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139